### PR TITLE
Synchronize adding a child task to a TaskGroup with parent task

### PR DESCRIPTION
**Pull request by rokhinip**
When a task is adding adding new children to a task group, we need to synchronize with the task status record lock of the parent task that has the task group, to prevent races with concurrent cancellation and escalation.

In the future, we'd probably want a lock for the TaskGroup itself and not have it be the group's parent task's status record lock but for now, this will get us the right functionality. 

rdar://86311782

### CODE REVIEW

**Review by rjmccall — APPROVED** [review]
LGTM

**Comment by ktoso** [comment]
Resolves rdar://86311782

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/TaskStatus.cpp [hunk 1]
```diff
@@ -468,6 +468,27 @@ swift_task_detachChildImpl(ChildTaskStatusRecord *record) {
   swift_task_removeStatusRecord(record);
 }
 
+SWIFT_CC(swift)
+static void swift_taskGroup_attachChildImpl(TaskGroup *group,
+                                            AsyncTask *child) {
+  SWIFT_TASK_DEBUG_LOG("attach child task = %p to group = %p", child, group);
```
Comment (ktoso) [inline stdlib/public/Concurrency/TaskStatus.cpp:474]: We'll double-log this now, maybe remove from here?
```suggestion
```

File: stdlib/public/CompatibilityOverride/CompatibilityOverrideConcurrency.def [hunk 2]
```diff
@@ -241,7 +241,7 @@ OVERRIDE_TASK_GROUP(taskGroup_initialize, void,
                     SWIFT_EXPORT_FROM(swift_Concurrency), SWIFT_CC(swift),
                     swift::, (TaskGroup *group, const Metadata *T), (group, T))
 
-OVERRIDE_TASK_GROUP(taskGroup_attachChild, void,
+OVERRIDE_TASK_STATUS(taskGroup_attachChild, void,
```
Comment (ktoso) [inline stdlib/public/CompatibilityOverride/CompatibilityOverrideConcurrency.def:244]: I _think_ this is fine but am not 100% sure, the overrides I think are for older deployments in theory... Let's double check with @DougGregor just to be sure if we can change this.

**Review by ktoso — APPROVED** [review]
LGTM, thanks for spotting this!

Feel free to merge once CI passes, removing the double-log can be done any time.

**Comment by ktoso** [comment]
Failure was the unrelated 
```
/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test/branch-main/swiftpm/Tests/WorkspaceTests/RegistryPackageContainerTests.swift:421:39: error: type 'FingerprintCheckingMode' has no member 'none'
20:09:38             fingerprintCheckingMode: .none,
20:09:38                                      ~^~~~
```
from swiftpm that we fixed; rebuilding

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by rokhinip** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
rdar://86311782

File: stdlib/public/CompatibilityOverride/CompatibilityOverrideConcurrency.def (hunk 2 above)
Comment (DougGregor) [reply stdlib/public/CompatibilityOverride/CompatibilityOverrideConcurrency.def:244]: Yeah, it's fine. It only affects what file stamps out the boilerplate here.

**Pull request merged**